storage/blockchain/src/service/read.rs

```diff
@@ -525,3 +530,60 @@ fn key_images_spent(env: &ConcreteEnv, key_images: HashSet<KeyImage>) -> Respons
         Some(Err(e)) => Err(e), // A database error occurred.
     }
 }
+
+/// [`BCReadRequest::CompactChainHistory`]
+fn compact_chain_history(env: &ConcreteEnv) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = env_inner.tx_ro()?;
+
+    let table_block_heights = env_inner.open_db_ro::<BlockHeights>(&tx_ro)?;
+    let table_block_infos = env_inner.open_db_ro::<BlockInfos>(&tx_ro)?;
+
+    let top_block_height = top_block_height(&table_block_heights)?;
+
+    let top_block_info = get_block_info(&top_block_height, &table_block_infos)?;
+    let cumulative_difficulty = combine_low_high_bits_to_u128(
+        top_block_info.cumulative_difficulty_low,
+        top_block_info.cumulative_difficulty_high,
+    );
+
+    /// The amount of top block IDs in the compact chain.
+    const INITIAL_BLOCKS: u64 = 11;
+
+    let mut block_ids = (0..)
+        .map(compact_history_index_to_height_offset::<INITIAL_BLOCKS>)
+        .map_while(|i| top_block_height.checked_sub(i))
+        .map(|height| Ok(get_block_info(&height, &table_block_infos)?.block_hash))
+        .collect::<Result<Vec<_>, RuntimeError>>()?;
+
+    if compact_history_genesis_not_included::<INITIAL_BLOCKS>(top_block_height) {
+        block_ids.push(get_block_info(&0, &table_block_infos)?.block_hash);
+    }
+
+    Ok(BCResponse::CompactChainHistory {
+        cumulative_difficulty,
+        block_ids,
+    })
+}
+
+/// [`BCReadRequest::FindFirstUnknown`]
+fn find_first_unknown(env: &ConcreteEnv, block_ids: Vec<BlockHash>) -> ResponseResult {
+    let env_inner = env.env_inner();
+    let tx_ro = env_inner.tx_ro()?;
+
+    let table_block_heights = env_inner.open_db_ro::<BlockHeights>(&tx_ro)?;
+
+    let mut last_height = 0;
+
+    for (i, block_id) in block_ids.into_iter().enumerate() {
+        match get_block_height(&block_id, &table_block_heights) {
+            Ok(height) => last_height = height,
+            Err(RuntimeError::KeyNotFound) => {
+                return Ok(BCResponse::FindFirstUnknown(Some((i, last_height + 1))))
+            }
+            Err(e) => return Err(e),
+        }
+    }
+
+    Ok(BCResponse::FindFirstUnknown(None))
+}
```

Comment:
Same here, what is `block_ids.len()` typically going to be?

Reply:
Max is 25,000 which now that I say it is a lot ... I'll change this to a binary search.